A CEGAR approach to compute objects from the value set analysis

lucasccordeiro opened the pull request:
This PR aims to implement a counterexample-guided abstraction refinement approach to compute objects from the value set analysis and insert them as assumptions in the code. 

The algorithm will work as follows:

```
1) Extract the possible objects that the pointer can point to via value set analysis (over-approximation).
2) Introduce this set of objects for each pointer as an assert statement and check each assertion incrementally.
3) If the assertion in step 2 holds, we turn it into an assume statement to help the k-induction proof rule.
4) If step 3 fails, we extract the counterexample and refine our specification. Go back to step 3.
5) We set a threshold if step 4 fails to refine the invariant.
```

lucasccordeiro committed:
Convert objects from the value set analysis

Here, we convert the objects that the pointer
points to via the value set analysis. In particular,
we convert these objects into an assume statement,
which defines the set of objects that the pointer
can point to.

lucasccordeiro committed:
Added regression tests for checking the value set

Added some regression tests to check whether the
objects obtained from the value set analysis
help our k-induction and incremental solving
to prove partial correctness.

lucasccordeiro committed:
Do not assign nondet value to pointers

If we add the objects obtained during the value set
analysis, we should not assign nondet values to pointers
since our symbolic execution will consider the NULL pointer
as one of those objects.

lucasccordeiro committed:
Added test cases to check for pointer assignment

These test cases stress out the need for not assiggning
nondet values for pointers, which make the inductive step
fail during the program verification.

lucasccordeiro committed:
added regression test with non-trivial invariant

Added test cases with non-trivial invariants to be checked
by our k-induction and incremental SMT.

lucasccordeiro commented:
Results of this PR over the `memory-safety` category of SV-COMP 2021.

```
Statistics:           3434 Files
  correct:            2797
    correct true:     1383
    correct false:    1414
  incorrect:            24
    incorrect true:      3
    incorrect false:    21
  unknown:             613
  Score:              3748 (max: 5227)
```

lucasccordeiro commented:
Results of the master branch over the `memory-safety` category of SV-COMP 2021.

```
Statistics:           3434 Files
  correct:            2795
    correct true:     1382
    correct false:    1413
  incorrect:            24
    incorrect true:      3
    incorrect false:    21
  unknown:             615
  Score:              3745 (max: 5227)
```

lucasccordeiro commented:
This PR can verify two additional programs (one TRUE and one FALSE) from the `memory-safety` category, thus improving our results from `3745` to `3748`.

lucasccordeiro commented:
The difference seems to be just a `noise`:

This PR solves this program `sll_length_check-2` in 289 seconds, while the master branch leads to a time-out (i.e., > 300 seconds).

lucasccordeiro commented:
I'm now running this PR over the entire SV-COMP 2021 suite.

rafaelsamenezes commented on src/goto-symex/symex_dereference.cpp:
```diff
@@ -52,7 +54,56 @@ void symex_dereference_statet::get_value_set(
+  // Here we obtain the set of objects via value set analysis
   state.value_set.get_value_set(expr, value_set);
+
+  default_message msg;
```
Is this needed?

lucasccordeiro replied:
Well spotted! Fixed.

rafaelsamenezes reviewed (approved):
LGTM

lucasccordeiro committed:
removed unnecessary message variable

mikhailramalho commented on src/goto-symex/symex_dereference.cpp:
```diff
@@ -52,7 +52,54 @@ void symex_dereference_statet::get_value_set(
+      for(auto const &i : value_set)
+      {
+        // note that the set of objects are encoded as object_descriptor
+        if(i->expr_id != expr2t::object_descriptor_id)
```
is_object_descriptor2t(i)

lucasccordeiro replied:
Fixed.

lucasccordeiro replied:
Fixed.

mikhailramalho commented on src/goto-symex/symex_dereference.cpp:
```diff
@@ -52,7 +52,54 @@ void symex_dereference_statet::get_value_set(
+      // we will accumulate the objects that the pointer points to
+      expr2tc or_accuml = gen_false_expr();
+      same_object2tc eq;
+
```
there is a simpler algorithm for this:

1. or_accuml = value_set[0]. 
2. for (size_t i = 1; i < value_set.size(); ++i) or_accuml = or2tc(or_accuml, value_set[i];

Then there is no need for the is_false guard after the loop.

lucasccordeiro replied:
@mikhailramalho: You're correct about the check for the `is_false_guard`, which is not needed. However, I don't understand your first suggestion.

Note that we want to create a logical equation such as `ptr points to object1 || ptr points to object2 || ... || ptr points to objectn`. However, you're suggesting something like `object1 || object2 || .. || objectn`.

Is my interpretation correct?

Note that we now consider the pointer offset when building the logical formula with the possible objects that the pointer can point to.

I'll now run this patch over the SV-COMP benchmarks and then compare with the current results of the master, which I already reported in this issue.

mikhailramalho replied:
No, what I meant is that you're creating the following formula: `false || ptr points to object1 || ptr points to object2 || ... || ptr points to object`, i.e., the `false` comes from `expr2tc or_accuml = gen_false_expr();`.

You don't need that, simply initialize `or_accuml` with the first element of the value set and loop from position 1 in the value set (you probably gonna need a while loop here instead of a for).

lucasccordeiro replied:
Oh I see. I'll take a look at it.

mikhailramalho reviewed (changes requested):
I still find it weird that we don't use the object's offset. Could you create some examples where a pointer can point to several positions in an array?

lucasccordeiro commented:
Results from the master branch:

```
Statistics:          15201 Files
  correct:            7701
    correct true:     4899
    correct false:    2802
  incorrect:            58
    incorrect true:     25
    incorrect false:    33
  unknown:            7442
  Score:             11272 (max: 24562)
```

Results from this PR:
```
Statistics:          15201 Files
  correct:            7191
    correct true:     4513
    correct false:    2678
  incorrect:            43
    incorrect true:     16
    incorrect false:    27
  unknown:            7967
  Score:             10760 (max: 24562)
```

lucasccordeiro committed:
added is_object_descriptor2t method

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
added regression tests with pointers to arrays

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
consider only object descriptors for value-set analysis

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
consider pointer offset

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
there is no need for the is_false guard

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
added regression test from SV-COMP 2021

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
initialize pointers with nondet if --no-pointer is enabled

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
Fixed or_accuml to store the number of objects

Note that the pointer could point to any of the
accumulated objects. However, if we have just
one element, our `or_accuml` should store just
that single element; otherwise, we will accumulate
all the other expressions.

lucasccordeiro committed:
Compare with NULL object

If we obtain a NULL object from the value-set analysis,
we should compare the current pointer with that NULL
object.

lucasccordeiro commented:
The experimental results over SV-COMP show that this PR reduced the number of both incorrect and correct results. It seems that if we add these assume statements, ESBMC becomes slower.

I'll check some particular cases to confirm my findings.

lucasccordeiro commented:
I have inspected the results. We've several cases failing due to:

```
Couldn't convert expression in unrecognised format
unknown
* type : signedbv
  * width : 32
```

lucasccordeiro committed:
Check whether the object offset is unknown.

If the object offset is unknown, we need to take care
of this specific case before doing the pointer arithmetic
with the object base address. For now, we don't add the
assumption.

lucasccordeiro committed:
added regression test for unknown case

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

mikhailramalho committed:
Reduce number of used variables

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

lucasccordeiro commented:
I started running this PR over the SV-COMP benchmarks in `FM10` (folder `value-set`).

lucasccordeiro commented:
The latest results of this PR:

```
Statistics:          15201 Files
  correct:            7662
    correct true:     4871
    correct false:    2791
  incorrect:            57
    incorrect true:     25
    incorrect false:    32
  unknown:            7482
  Score:             11221 (max: 24562)
```

Results from the master branch:

```
Statistics:          15201 Files
  correct:            7701
    correct true:     4899
    correct false:    2802
  incorrect:            58
    incorrect true:     25
    incorrect false:    33
  unknown:            7442
  Score:             11272 (max: 24562)
```

lucasccordeiro commented:
The benchmark `/linux-3.14_complex_emg_linux-alloc-spinlock_drivers-net-ethernet-myricom-myri10ge-myri10ge.cil` produces a timeout with this PR, while the master branch produces `incorrect false` in 45 seconds (CPU time).

mikhailramalho commented:
Does it get stuck in solving or is it something else?

Em qua., 28 de jul. de 2021 às 15:33, Lucas Cordeiro <
***@***.***> escreveu:

> The benchmark
> /linux-3.14_complex_emg_linux-alloc-spinlock_drivers-net-ethernet-myricom-myri10ge-myri10ge.cil
> produces a timeout with this PR, while the master branch produces incorrect
> false in 45 seconds (CPU time).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/468#issuecomment-888566789>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH3ZABDCCUWSAKUZFILT2BLQPANCNFSM5AYTRLEQ>
> .
>


-- 

Mikhail Ramalho.

lucasccordeiro commented:
> Does it get stuck in solving or is it something else?

It keeps unfolding the program and checking the resulting formula.

lucasccordeiro commented:
As the next step, we could create another PR to introduce and check these invariants during the inductive step only.

mikhailramalho commented:
Weird, does it means that the invariants are unsound?

On Wed, 28 Jul 2021, 15:41 Lucas Cordeiro, ***@***.***> wrote:

> Does it get stuck in solving or is it something else?
>
> It keeps unfolding the program and checking the resulting formula.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/468#issuecomment-888571438>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH45BZ2JFHPGPAIAQBTT2BMPFANCNFSM5AYTRLEQ>
> .
>

lucasccordeiro commented:
What do you mean by unsound?

The master branch produces an incorrect false while this PR produces a time-out.


> On 28 Jul 2021, at 21:32, Mikhail R. Gadelha ***@***.***> wrote:
> 
> ﻿
> Weird, does it means that the invariants are unsound?
> 
> On Wed, 28 Jul 2021, 15:41 Lucas Cordeiro, ***@***.***> wrote:
> 
> > Does it get stuck in solving or is it something else?
> >
> > It keeps unfolding the program and checking the resulting formula.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/esbmc/esbmc/pull/468#issuecomment-888571438>, or
> > unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/AAKEJH45BZ2JFHPGPAIAQBTT2BMPFANCNFSM5AYTRLEQ>
> > .
> >
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
